From 193708997d8b8e97e07c45ce1ce66feb6b801302 Mon Sep 17 00:00:00 2001
From: Michael Webster <miketwebster@gmail.com>
Date: Mon, 15 Dec 2025 15:22:00 -0500
Subject: [PATCH] templates: Improve invalid XDG_TEMPLATES_DIR handling.

When xdg-user-dirs-update is executed, entries with missing
locations are set to $HOME.

The nemo_should_use_templates_directory() already guards against
trying to load templates from that location, but our new prefs
Template page wasn't allowing for this (and would attempt to
recurse every file in the user's home directory to add to its
template list).

Create the Templates folder and update XDG dirs *only* if the
user attempts to add templates or open the templates folder
directly, otherwise continue ignoring the missing or mis-
configured directory.

Fixes #3652.
---
 libnemo-private/nemo-file-utilities.c | 83 ++++++++++++++++++---------
 libnemo-private/nemo-file-utilities.h |  2 +-
 src/nemo-interesting-folder-bar.c     | 16 +++---
 src/nemo-template-config-widget.c     | 26 ++++++---
 src/nemo-window-menus.c               |  1 +
 5 files changed, 84 insertions(+), 44 deletions(-)

diff --git a/libnemo-private/nemo-file-utilities.c b/libnemo-private/nemo-file-utilities.c
index 033ff37a8..07d7fa1d0 100644
--- a/libnemo-private/nemo-file-utilities.c
+++ b/libnemo-private/nemo-file-utilities.c
@@ -51,6 +51,7 @@
 static void update_xdg_dir_cache (void);
 static void schedule_user_dirs_changed (void);
 static void desktop_dir_changed (void);
+static void update_xdg_user_dir (const char *type, const char *path);
 static GFile *nemo_find_file_insensitive_next (GFile *parent, const gchar *name);
 
 char *
@@ -305,6 +306,32 @@ parse_xdg_dirs (const char *config_file)
 static XdgDirEntry *cached_xdg_dirs = NULL;
 static GFileMonitor *cached_xdg_dirs_monitor = NULL;
 
+static void
+update_xdg_user_dir (const char *type, const char *path)
+{
+    char *argv[5];
+    int i;
+
+    i = 0;
+    argv[i++] = (char *)"xdg-user-dirs-update";
+    argv[i++] = (char *)"--set";
+    argv[i++] = (char *)type;
+    argv[i++] = (char *)path;
+    argv[i++] = NULL;
+
+    /* We do this sync, to avoid possible race-conditions
+       if multiple dirs change at the same time. Its
+       blocking the main thread, but these updates should
+       be very rare and very fast. */
+    g_spawn_sync (NULL,
+                  argv, NULL,
+                  G_SPAWN_SEARCH_PATH |
+                  G_SPAWN_STDOUT_TO_DEV_NULL |
+                  G_SPAWN_STDERR_TO_DEV_NULL,
+                  NULL, NULL,
+                  NULL, NULL, NULL, NULL);
+}
+
 static void
 xdg_dir_changed (NemoFile *file,
 		 XdgDirEntry *dir)
@@ -318,30 +345,10 @@ xdg_dir_changed (NemoFile *file,
 		path = g_file_get_path (location);
 
 		if (path) {
-			char *argv[5];
-			int i;
-
 			g_free (dir->path);
 			dir->path = path;
 
-			i = 0;
-			argv[i++] = (char *)"xdg-user-dirs-update";
-			argv[i++] = (char *)"--set";
-			argv[i++] = dir->type;
-			argv[i++] = dir->path;
-			argv[i++] = NULL;
-
-			/* We do this sync, to avoid possible race-conditions
-			   if multiple dirs change at the same time. Its
-			   blocking the main thread, but these updates should
-			   be very rare and very fast. */
-			g_spawn_sync (NULL,
-				      argv, NULL,
-				      G_SPAWN_SEARCH_PATH |
-				      G_SPAWN_STDOUT_TO_DEV_NULL |
-				      G_SPAWN_STDERR_TO_DEV_NULL,
-				      NULL, NULL,
-				      NULL, NULL, NULL, NULL);
+			update_xdg_user_dir (dir->type, dir->path);
 			g_reload_user_special_dirs_cache ();
 			schedule_user_dirs_changed ();
 			desktop_dir_changed ();
@@ -608,15 +615,35 @@ nemo_get_templates_directory (void)
 }
 
 void
-nemo_create_templates_directory (void)
+nemo_ensure_valid_templates_directory (void)
 {
-	char *dir;
+    /* This is called only at points where the user would expect the Templates
+     * directory to exist:
+     * - Template prefs, adding a template via DND or the "New" button.
+     * - Template prefs - the Folder icon.
+     * - MenuBar->Go->Templates - navigating directly to the Templates dir.
+     *
+     * Otherwise, we're fine without it, if the user never intends to use templates.
+     */
+    g_autofree gchar *templates_dir = NULL;
+    const char *home_dir;
 
-	dir = nemo_get_templates_directory ();
-	if (!g_file_test (dir, G_FILE_TEST_EXISTS)) {
-		g_mkdir (dir, DEFAULT_NEMO_DIRECTORY_MODE);
-	}
-	g_free (dir);
+    templates_dir = nemo_get_xdg_dir ("TEMPLATES");
+    home_dir = g_get_home_dir ();
+
+    if (g_strcmp0 (templates_dir, home_dir) == 0) {
+        // XDG_TEMPLATES_DIR is $HOME/, fix it.
+        g_autofree gchar *fixed_templates_dir = NULL;
+        fixed_templates_dir = g_build_filename (home_dir, "Templates", NULL);
+        g_mkdir (fixed_templates_dir, DEFAULT_NEMO_DIRECTORY_MODE);
+
+        update_xdg_user_dir ("TEMPLATES", fixed_templates_dir);
+        g_reload_user_special_dirs_cache ();
+        update_xdg_dir_cache ();
+    } else {
+        // XDG_TEMPLATES_DIR is reasonable, let's make sure it exists.
+        g_mkdir (templates_dir, DEFAULT_NEMO_DIRECTORY_MODE);
+    }
 }
 
 char *
diff --git a/libnemo-private/nemo-file-utilities.h b/libnemo-private/nemo-file-utilities.h
index 5862acda0..504660c5a 100644
--- a/libnemo-private/nemo-file-utilities.h
+++ b/libnemo-private/nemo-file-utilities.h
@@ -57,8 +57,8 @@ char *   nemo_get_gmc_desktop_directory          (void);
 
 gboolean nemo_should_use_templates_directory     (void);
 char *   nemo_get_templates_directory            (void);
+void     nemo_ensure_valid_templates_directory   (void);
 char *   nemo_get_templates_directory_uri        (void);
-void     nemo_create_templates_directory         (void);
 
 char *   nemo_get_searches_directory             (void);
 
diff --git a/src/nemo-interesting-folder-bar.c b/src/nemo-interesting-folder-bar.c
index dcaf99583..6fc2744eb 100644
--- a/src/nemo-interesting-folder-bar.c
+++ b/src/nemo-interesting-folder-bar.c
@@ -236,14 +236,16 @@ nemo_interesting_folder_bar_new_for_location (NemoView *view, GFile *location)
     }
 
     if (type == TYPE_NONE_FOLDER) {
-        path = nemo_get_templates_directory ();
-        tmp_loc = g_file_new_for_path (path);
-        g_free (path);
-
-        if (g_file_equal (location, tmp_loc)) {
-            type = TYPE_TEMPLATES_FOLDER;
+        if (nemo_should_use_templates_directory ()) {
+            path = nemo_get_templates_directory ();
+            tmp_loc = g_file_new_for_path (path);
+            g_free (path);
+
+            if (g_file_equal (location, tmp_loc)) {
+                type = TYPE_TEMPLATES_FOLDER;
+            }
+            g_object_unref (tmp_loc);
         }
-        g_object_unref (tmp_loc);
     }
 
     return type == TYPE_NONE_FOLDER ? NULL : nemo_interesting_folder_bar_new (view, type);
diff --git a/src/nemo-template-config-widget.c b/src/nemo-template-config-widget.c
index 6659aaa04..85c966796 100644
--- a/src/nemo-template-config-widget.c
+++ b/src/nemo-template-config-widget.c
@@ -221,9 +221,11 @@ refresh_widget (NemoTemplateConfigWidget *widget)
 
     gchar *path = NULL;
 
-    path = nemo_get_templates_directory ();
-    populate_from_directory (widget, path);
-    g_clear_pointer (&path, g_free);
+    if (nemo_should_use_templates_directory ()) {
+        path = nemo_get_templates_directory ();
+        populate_from_directory (widget, path);
+        g_clear_pointer (&path, g_free);
+    }
 
     if (widget->templates == NULL) {
         GtkWidget *empty_label = gtk_label_new (NULL);
@@ -362,7 +364,8 @@ start_renaming_selected_row (NemoTemplateConfigWidget *widget, GtkWidget *row)
 }
 
 static void
-import_template (const gchar *filename)
+import_template (NemoTemplateConfigWidget *widget,
+                 const gchar              *filename)
 {
     GFile *source, *dest;
     gchar *basename, *template_dir, *dest_path;
@@ -385,6 +388,7 @@ import_template (const gchar *filename)
     }
 
     basename = g_file_get_basename (source);
+    nemo_ensure_valid_templates_directory ();
     template_dir = nemo_get_templates_directory ();
     dest_path = g_build_filename (template_dir, basename, NULL);
     dest = g_file_new_for_path (dest_path);
@@ -396,6 +400,10 @@ import_template (const gchar *filename)
     g_free (dest_path);
     g_object_unref (source);
     g_object_unref (dest);
+
+    if (g_list_length (widget->dir_monitors) == 0) {
+        refresh_widget (widget);
+    }
 }
 
 static gboolean
@@ -436,7 +444,7 @@ on_new_template_clicked (GtkWidget *button, gpointer user_data)
         gchar *selected_filename = gtk_file_chooser_get_filename (file_chooser);
 
         if (selected_filename != NULL) {
-            import_template (selected_filename);
+            import_template (widget, selected_filename);
         }
 
         g_free (selected_filename);
@@ -485,6 +493,7 @@ static void
 on_open_folder_clicked (GtkWidget *button, NemoTemplateConfigWidget *widget)
 {
     gchar *path = NULL;
+    nemo_ensure_valid_templates_directory ();
     path = nemo_get_templates_directory ();
     GFile *location = g_file_new_for_path (path);
 
@@ -507,6 +516,7 @@ on_drag_data_received(GtkWidget *widget,
                       guint time,
                       gpointer user_data)
 {
+    NemoTemplateConfigWidget *self = NEMO_TEMPLATE_CONFIG_WIDGET (user_data);
     gboolean success = FALSE;
 
     if (data != NULL && gtk_selection_data_get_length(data) >= 0) {
@@ -517,14 +527,14 @@ on_drag_data_received(GtkWidget *widget,
 
             if (uris != NULL && uris[0] != NULL) {
                 for (int i = 0; uris[i] != NULL; i++) {
-                    import_template (uris[i]);
+                    import_template (self, uris[i]);
                 }
 
                 g_strfreev(uris);
                 success = TRUE;
             }
         } else if (target_type == TARGET_TEXT_PLAIN) {
-            import_template ((const gchar *) raw_data);
+            import_template (self, (const gchar *) raw_data);
             success = TRUE;
         }
     }
@@ -655,7 +665,7 @@ nemo_template_config_widget_init (NemoTemplateConfigWidget *self)
                       GDK_ACTION_COPY);
 
     g_signal_connect(NEMO_CONFIG_BASE_WIDGET (self)->listbox, "drag-data-received",
-                    G_CALLBACK(on_drag_data_received), label);
+                    G_CALLBACK(on_drag_data_received), self);
     g_signal_connect(NEMO_CONFIG_BASE_WIDGET (self)->listbox, "drag-motion",
                     G_CALLBACK(on_drag_motion), NULL);
 
diff --git a/src/nemo-window-menus.c b/src/nemo-window-menus.c
index fd374b3df..46f10b8e6 100644
--- a/src/nemo-window-menus.c
+++ b/src/nemo-window-menus.c
@@ -200,6 +200,7 @@ action_go_to_templates_callback (GtkAction *action,
 	window = NEMO_WINDOW (user_data);
 	slot = nemo_window_get_active_slot (window);
 
+	nemo_ensure_valid_templates_directory ();
 	path = nemo_get_templates_directory ();
 	location = g_file_new_for_path (path);
 	g_free (path);
